### PR TITLE
Feat/dedup-lease (part 1)

### PR DESCRIPTION
**Pull request by iGxnon**
Signed-off-by: iGxnon <igxnon@gmail.com>

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

    Add client id lease mechanism, this is the part 1 PR.

* what changes does this pull request make?

    Implement lease server & lease keep alive connection.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No.

### CODE REVIEW

File: curp/src/rpc/connect.rs [hunk 1]
```diff
@@ -583,9 +636,37 @@ where
...
+    /// Keep send lease keep alive to server and mutate the client id
+    async fn lease_keep_alive(&self, _client_id: Arc<RwLock<u64>>) -> CurpError {
+        unreachable!("cannot invoke lease_keep_alive in bypassed connect")
+    }
+}
+
+/// Generate heartbeat stream
+fn heartbeat_stream(client_id: Arc<RwLock<u64>>) -> impl Stream<Item = LeaseKeepAliveMsg> {
+    /// Keep alive interval
+    const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(1);
```
Comment (Phoenix500526) [inline curp/src/rpc/connect.rs:653]: Shall we make this interval configurable?

Comment (iGxnon) [reply curp/src/rpc/connect.rs:695]: Fixed in #617

File: curp/src/server/raw_curp/mod.rs [hunk 2]
```diff
@@ -541,6 +550,22 @@ impl<C: Command, RC: RoleChange> RawCurp<C, RC> {
...
+        }
+        self.ctx.lm.map_write(|mut lm_w| {
+            if lm_w.check_alive(client_id) {
+                lm_w.renew(client_id);
+                None
+            } else {
+                lm_w.revoke(client_id);
+                Some(lm_w.grant())
+            }
+        })
```
Comment (themanforfree) [inline curp/src/server/raw_curp/mod.rs:566]: All operations in this method require a write lock of lease_manager, and I recommend getting a write lock directly at the beginning of the method to reduce nesting

File: curp/src/server/lease_manager.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
+use std::cmp::Reverse;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::Duration;
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply curp/src/server/lease_manager.rs]: done

File: curp/src/server/lease_manager.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        } else {
+            false
+        }
+    }
+
+    /// Generate a new client id and grant a lease
+    pub(crate) fn grant(&mut self) -> u64 {
+        let client_id: u64 = rand::random();
+        let expiry = Instant::now().add(DEFAULT_LEASE_TTL);
+        let _ig = self.expiry_queue.push(client_id, Reverse(expiry));
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply curp/src/server/lease_manager.rs]: done

File: curp/src/server/raw_curp/mod.rs (hunk 2 above)
Comment (iGxnon) [reply curp/src/server/raw_curp/mod.rs]: done

File: curp/src/rpc/connect.rs [hunk 5]
```diff
@@ -204,6 +204,9 @@ pub(crate) trait ConnectApi: Send + Sync + 'static {
         request: MoveLeaderRequest,
         timeout: Duration,
     ) -> Result<tonic::Response<MoveLeaderResponse>, CurpError>;
+
+    /// Keep send lease keep alive to server and mutate the client id
+    async fn lease_keep_alive(&self, client_id: Arc<RwLock<u64>>) -> CurpError;
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply curp/src/rpc/connect.rs]: done, refactored to AtomicU64

**Review by themanforfree — APPROVED** [review]
LGTM

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**